File: 2018/day-1/node/main_pt2.js
var fs = require("fs");

console.log("-----------------------------");
console.log("Advent of Code - Day 1 - Node");

function findDuplicateFrequency(filename) {
  var freq = 0;
  const visitedFrequencies = new Set([freq]);

  var lines = fs.readFileSync(filename, "utf8").split("\n");

  while (true) {
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      freq += parseInt(line);

      // Check if frequency already visited
      if (visitedFrequencies.has(freq)) {
        return freq;
      }

      // If not, add it to list
      visitedFrequencies.add(freq);
    }
  }
}

freq = findDuplicateFrequency("../input.txt");
console.log("First duplicate frequency is:", freq);

File: 2018/day-1/go/main_pt2.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

func findDuplicateFrequency(filename string) (int, error) {

	b, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}
	s := string(b)
	lines := strings.Split(s, "\n")

	visitedFrequencies := make(map[int]struct{})

	freq := 0
	for {
		for _, line := range lines {
			i, err := strconv.Atoi(line)
			if err != nil {
				log.Fatal("Input not valid", err)
			}
			freq += i

			// Check if frequency has been visited before
			if _, ok := visitedFrequencies[freq]; ok {
				return freq, nil
			}

			// If not, add it to the list
			visitedFrequencies[freq] = struct{}{}
		}
	}
}

func main() {
	fmt.Println("---------------------------")
	fmt.Println("Advent of Code - Day 2 - Go")

	duplicatedFrequency, err := findDuplicateFrequency("../input.txt")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("First duplicate frequency is:", duplicatedFrequency)
}
